### ISSUE TITLE
Fail of customized csr matrix load after update to 0.8.1

### ISSUE THREAD

**Issue by joachimwolff** [mirnylab/cooler#145]
Hi,

to load a cooler matrix into a scipy csr matrix I use my own code to have a faster performance and less memory usage compared to `cooler_file.matrix(balance=False, sparse=True, as_pixels=False).tocsr()`. What I do is to load the data from the pandas data frame in chunks per bins  and build it to csr matrix together:

```python {.line-numbers}
cooler_file = cooler.Cooler(self.matrixFileName)
matrixDataFrame = cooler_file.matrix(balance=False, sparse=True, as_pixels=True)
used_dtype = np.int32
if np.iinfo(np.int32).max < cooler_file.info['nbins']:
    used_dtype = np.int64
count_dtype = matrixDataFrame[0]['count'].dtype
data = np.empty(cooler_file.info['nnz'], dtype=count_dtype)
instances = np.empty(cooler_file.info['nnz'], dtype=used_dtype)
features = np.empty(cooler_file.info['nnz'], dtype=used_dtype)
i = 0
size = cooler_file.info['nbins'] // 32
if size == 0:
    size = 1
start_pos = 0

while i < cooler_file.info['nbins']:
    _data = matrixDataFrame[i:i + size]['count'].values.astype(count_dtype).T
    _instances = matrixDataFrame[i:i + size]['bin1_id'].values.astype(used_dtype).T
     _features = matrixDataFrame[i:i + size]['bin2_id'].values.astype(used_dtype).T

    data[start_pos:start_pos + len(_data)] = _data
    instances[start_pos:start_pos + len(_instances)] = _instances
    features[start_pos:start_pos + len(_features)] = _features
    start_pos += len(_features)
    i += size
matrix = csr_matrix((data, (instances, features)), shape=(cooler_file.info['nbins'], cooler_file.info['nbins']), dtype=count_dtype)
            
```

This code works perfectly with cooler 0.7.11, however it crashes after an update to 0.8.1.
The error is that the number of elements from `cooler_file.info['nnz']` does not match the number of elements received via the iteration of `matrixDataFrame[i:i + size]['count'].` anymore and therefore the insertion to `data[start_pos:start_pos + len(_data)] = _data` fails in the last iteration. 

Did you changed anywhere something which is not documented or I have overseen? Any help would be appreciated.

Thanks a lot.

Best,

Joachim

**Comment by nvictus** [comment]
Hi Joachim,

A bug was introduced into the "matrix-as-pixels" selector in 0.8.1 causing the output to return as the transpose in certain cases. Apologies. It will be fixed asap.

I have some questions about your usage pattern to better understand how you use the matrix-as-pixels selector.

* Do you ever use it to extract sub-matrices or simply to populate the entire dataset into a coo_matrix?
* Do you keep the upper triangular representation, or do you *duplex* the data afterwards, as the numpy as scipy selectors do?

Also, I noticed one thing:
> ```python
>     _data = matrixDataFrame[i:i + size]['count'].values.astype(count_dtype).T
>     _instances = matrixDataFrame[i:i + size]['bin1_id'].values.astype(used_dtype).T
>      _features = matrixDataFrame[i:i + size]['bin2_id'].values.astype(used_dtype).T
> ```

`matrixDataFrame` is a selector, where row slices trigger querying, so the above reads all 3 columns three times. Switching the order of the accession (e.g. `matrixDataFrame['count'][i:i + size]`) will only read a single column at a time, which is probably what you intended. Alternatively, you can do:

```python
# reads all columns into a data frame
result = matrixDataFrame[i:i + size]
# I don't think the .T is necessary since Series are 1D
_data = result['count'].values.astype(count_dtype) 
_instances = result['bin1_id'].values.astype(used_dtype)
_features = result['bin2_id'].values.astype(used_dtype)
```

Finally, if you are loading the entire dataset into a scipy `coo_matrix` (no sub-matrix selection), then the 1D pixel table selector will be more convenient for you, because you can control precisely how many pixels you load in memory at a time, rather than iterating over matrix rows. e.g.:

```python
edges = np.linspace(0, cooler_file.info['nnz'], 32, dtype=int)
selector = cooler_file.pixels()
for lo, hi in zip(edges[:-1], edges[1:]):
    result = selector[lo:hi]
    instances[lo:hi] = result['bin1_id'].values.astype(used_dtype)
    features[lo:hi] = result['bin2_id'].values.astype(used_dtype)
    data[lo:hi] = result['count'].values.astype(count_dtype)

matrix = csr_matrix(
    (data, (instances, features)), 
    shape=(cooler_file.info['nbins'], cooler_file.info['nbins']),
    dtype=count_dtype)
```

Best,

Nezar

**Comment by joachimwolff** [comment]
Hi Nezar,

thanks for your reply and quick bug fixing. I can confirm the bug is fixed with 0.8.2.

> Do you ever use it to extract sub-matrices or simply to populate the entire dataset into a coo_matrix?

I just populate the full matrix with it. I guess the problematic function is the `tocsr()` from numpy.

>Do you keep the upper triangular representation, or do you duplex the data afterwards, as the numpy as scipy selectors do?

Usually just the upper triangle representation. However, there are parts in the code where a symmetric matrix representation is used. But this is old code that works and I will not touch it :)

>matrixDataFrame is a selector, where row slices trigger querying, so the above reads all 3 columns three times. Switching the order of the accession (e.g. matrixDataFrame['count'][i:i + size]) will only read a single column at a time, which is probably what you intended. 

Thanks for the advice, I didn't know this.

> Finally, if you are loading the entire dataset into a scipy coo_matrix (no sub-matrix selection), then the 1D pixel table selector will be more convenient for you, because you can control precisely how many pixels you load in memory at a time, rather than iterating over matrix rows. e.g.:

I will try this, everything which helps to the reduce the amount of memory usage is welcome.

Thanks again for fixing the bugs.

Best,

Joachim

**Comment by nvictus** [comment]
Thanks for the feedback! Much appreciated.
